jni/src/faiss_index_service.cpp

```diff
@@ -57,6 +57,103 @@ void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    // Add extra parameters that cant be configured with the index factory
+    SetExtraParameters<faiss::Index, faiss::IndexIVF, faiss::IndexHNSW>(jniUtil, env, parameters, indexWriter.get());
+
+    // Check that the index does not need to be trained
+    if(!indexWriter->is_trained) {
+        throw std::runtime_error("Index is not trained");
+    }
+
+    // Add vectors
+    std::unique_ptr<faiss::IndexIDMap> idMap (faissMethods->indexIdMap(indexWriter.release()));
+
+    // Check to see if the current index is HNSW
+    faiss::IndexHNSW * hnsw = dynamic_cast<faiss::IndexHNSW *>(idMap->index);
+
+    if(hnsw != NULL) {
+        // Check to see if the HNSW storage is IndexFlat
+        faiss::IndexFlat * storage = dynamic_cast<faiss::IndexFlat *>(hnsw->storage);
+        if(storage != NULL) {
+            // Allocate enough memory for all of the vectors we plan on inserting
+            // We do this to avoid unnecessary memory allocations during insert
+            storage->codes.reserve(dim * numVectors * 4);
+        }
+    }
+
+    return reinterpret_cast<jlong>(idMap.release());
+}
+
+void IndexService::insertToIndex(
+        int dim,
+        int numIds,
+        int threadCount,
+        int64_t vectorsAddress,
+        std::vector<int64_t> & ids,
+        jlong idMapAddress
+    ) {
+    // Read vectors from memory address (unique ptr since we want to remove from memory after use)
+    std::vector<float> * inputVectors = reinterpret_cast<std::vector<float>*>(vectorsAddress);
+
+    // The number of vectors can be int here because a lucene segment number of total docs never crosses INT_MAX value
+    int numVectors = (int) (inputVectors->size() / (uint64_t) dim);
+    if(numVectors == 0) {
+        throw std::runtime_error("Number of vectors cannot be 0");
+    }
+
+    if (numIds != numVectors) {
+        throw std::runtime_error("Number of IDs does not match number of vectors");
+    }
+
+    // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    faiss::IndexIDMap * idMap = reinterpret_cast<faiss::IndexIDMap *> (idMapAddress);
+
+    // Add vectors
+    idMap->add_with_ids(numVectors, inputVectors->data(), ids.data());
+}
+
+void IndexService::writeIndex(
+        int threadCount,
+        std::string indexPath,
+        jlong idMapAddress
+    ) {
+    // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
+    }
+
+    faiss::IndexIDMap * idMap = reinterpret_cast<faiss::IndexIDMap *> (idMapAddress);
+
+    // Write the index to disk
+    faissMethods->writeIndex(idMap, indexPath.c_str());
+
+    // Free the memory used by the index
+    delete idMap;
```

Comment:
One concern here is the case where we fail to release this index memory due to some exception during indexing process. 

I think this delete part should have its own method and we need to put the entire indexing process under try block in java side and call this delete method in finally block.

Reply:
This is a good catch. I think that doing the same smart pointer trick should work here though.

Reply:
Yes , Smart Pointer releases memory automatically  when it goes out of scope, we don't have to write delete idmap, it's automatically taken care.
https://fastercapital.com/content/Smart-Pointers-in-C----Managing-Memory-Safely.html#:~:text=Instead%20of%20returning%20a%20raw,overall%20safety%20of%20your%20code.

Reply:
There is an issue though with needing to release the memory of the underlying index as well in the case of an exception. Going to think about the best way to do that.

Reply:
> This is a good catch. I think that doing the same smart pointer trick should work here though.
Smart pointer is one thing to handle exception inside this method itself. How about the case where we are missing to call this method from java side? Are we making sure that this method is called once the init index is called?

Reply:
Ah, this could be an issue. I'll wrap the insertion method inside of a try catch block that will free the index.

Reply:
Pushed